keps/sig-network/3458-dynamically-re-configure-lb-node-set/README.md

```diff
@@ -0,0 +1,787 @@
+  between a single replica of the service proxy experiencing issues
+  watching/receiving endpoints updates, vs. all of the replicas. When all of the
+  replicas are experiencing read issues, likely, the best thing to do is not to
+  simultaneously fail all health checks on all nodes, causing a general
+  cluster-wide ingress outage for all services.
+C) that its external network related dependencies are fine: that it can both read
+  / write to them as intended.
+
+#### Describing what "healthy" currently is for Kube-proxy
+
+This section describes how Kube-proxy handles all above cases currently.
+
+A) Kube-proxy does not handle the first case. If Kube-proxy restarts it will stop
+  answering to the health check probes while it's not running, this is because
+  the health check server runs as a go-routine in the Kube-proxy process.  
+
+B) Kube-proxy currently validates that whatever service/endpoints event it
+  receives are also processed within a specific timeout, which is configurable.
+  Any event received will set the field `oldestPendingQueued` on its health
+  check server implementation. The value will correspond to the time the event
+  was observed and once that event is processed successfully: the same field is
+  unset to a zero time value. The health check server performs two validations
+  and reports green if: that field is the zero value, or if the field is not the
+  zero value but within the timeout range. This currently handles the "general
+  API server read issue" case mentioned above, whereby if all Kube-proxies
+  encounter issues reading from the API server: the state of the healthz
+  endpoints will indicate green from all of them, given that the last event
+  received was processed successfully. It does not handle the case where one / a
+  group of Kube-proxy replicas can't read from the API server, since these
+  replicas will continue reporting green, but probably should not be. It could
+  be that Kube-proxy is using invalid credentials when communicating with the
+  API server and that the node is not actually experiencing a loss of API server
+  connectivity. This case could have pods move around without Kube-proxy being
+  able to update its rules. In the worse case: this could have Kube-proxy point
+  load balancer traffic to old pod IPs which have been recycled and are now
+  being used by new pods - leading to security concerns.
+
+C) The third case is partially covered by the validation performed Kube-proxy
+  [when it initializes](1) - which validates that the configuration for its
+  dependent networking sub-system is correctly configured. Its "runtime" checks
+  are covered by it's current health check, since, if it experiences any issues
+  updating the networking rules: it won't set `oldestPendingQueued` to the zero
+  time value and hence fail the health check.
+
+It is important to mention that all of these checks are currently being
```

Comment:
"unschedulable" doesn't mean that though. Simply doing `kubectl cordon node` will have that effect too, and there can be a variety of reasons that a user might chose to execute that command.

Reply:
Do we have a better signal?  We have cluster-autoscaler's bespoke taint, but that is far from 100% coverage.

In my experience, `cordon` is usually followed by `drain` and then decommission.  There is no "draining" indicator (I think?) so `unschedulable` is as close as we can get.  If that flips back to false (== "schedulable") then we would add the node back to the LB set.

Reply:
Not sure you remember this: https://github.com/kubernetes/kubernetes/pull/111691? We already tried using the "schedulable" field as an indicator for including/removing the node from the LB set. That's not right and will lead to bugs. 

> If we want to enable connection draining, we probably need to intersect all of this with some other signal which indicates that the node is "probably going away". 

Does this matter for simply removing the taint from the service controller sync loop, i.e the first KEP on our list?

Reply:
It should not matter for the first KEP, right?

I didn't mean to use `unschedulable` for the add/remove  - I meant for the on-node HC.

Today a typical node EOL looks like:
* kubectl drain
  * set unschedulable
* service controller removes the node from LB sets, breaking all connections
* drain pods
* delete node

After your cleanups:
* kubectl drain
  * set unschedulable
* drain pods
* delete node
* service controller removes the node from LB sets, breaking all connections (may have been done by the cloud in the background)

More ideally:
* kubectl drain
  * set unschedulable
* LB HCs start to fail, no new connections
* drain pods
* LB connections die off naturally
* delete node
* service controller removes the node from LB sets, breaking all connections (may have been done by the cloud in the background)

Right?

Reply:
Ah, I understand better now and I agree!

> It should not matter for the first KEP, right?

Agreed. I will add it this to the "kube-proxy better reporting health" KEP

Reply:
That being said, https://github.com/kubernetes/kubernetes/pull/90823 did mention:

> This caused a production outage when nodes in an autoscaling group were marked unschedulable so the autoscaler would gracefully terminate them - instead the ingress controllers on those nodes were immediately removed from the LB pool and no **new** ingress requests could be accepted.

I am not sure it makes sense to overthink the semantics of a written description on a 3 years old issue, but if "accepting new requests" really is primordial to ingress controllers running on cordoned nodes which are about to be deleted, then it might not be a good idea.

Reply:
Googd point for consideration in KEP2 :)